### ISSUE TITLE
Pre-release testing 

### ISSUE THREAD

**Issue by rarkins** [renovateapp/renovate#1453]
Placeholder feature request for discussion with @alexeagle and anyone else interested 

Idea: automate some form of pre-release feedback loop between upstream and downstream dependencies. Downstream dependents could opt in to getting branches and maybe even PRs for unstable releases published by upstream maintainers. Could be major pre-releases (eg 4.0.0-alpha.1) but also even minor or patch ones such as 3.5.0-pre.1 if the author is concerned there was a lot of change and “maybe it broke somebody”. By opting into this, downstream dependents get the benefit of hopefully less dependency updates, assuming upstream authors pay attention to the results. 

Like most other Renovate features, this could be enabled/disabled in any way from per repository down to per package. 

Next question is how to give feedback to upstream packages. If they use Renovate themselves they could opt into this feature for full functionality, eg maybe an Issue could be created per-release with stats. Alternatively there’s nothing stopping us providing public stats for all public upstream packages regardless of opt-in.

Note: this pretty much needs the app to be feasible, as there’s not really any easy way to share information back and forth with self hosted instances until one day eventually exposing an API etc.

**Comment by alexeagle** [comment]
Here's an example use-case:
Angular has the main monorepo (kind of an oxymoron...) and some other tooling that lives in other repos, notably the Angular CLI isn't in the monorepo, but also things like build toolchain in Bazel.

Right now we don't have a good way to discover when a change in one repo will break once it's integrated with the other repos. One way to do this is to have "floating" dependencies, so that every time CI runs, it might pick up the new version. But then the breakage isn't related to anything that changed in the repo that's now broken, instead it goes red randomly, and worse all the PRs at that time are also red for a reason outside the control of the PR author.

The idea here is to use renovate so that integration breakages appear as red PRs from renovate when it tries to update the Angular snapshot build, for example.

**Comment by rarkins** [comment]
@alexeagle so in the angular case you’d like to opt in your downstream angular packages - and anyone else from the community who wishes - to continuously test against `master` branch of angular repos, so you can be alerted of breakages? If so I assume you follow the principle of “master should always be releasable”? Otherwise there’d be lots of alerts if every time it’s not. 

Downstream packages can opt-in to continuously testing against upstream builds.

Upstream packages could configure a list of downstream repositories they trust enough to trigger upstream alerts if they break because of a master build - the canaries in the coal mine, so to speak.

**Comment by destroyerofbuilds** [comment]
CC @fenduru 

We are in a similar situation where we publish pre-releases of core libraries, but we have few options for comprehensive regression testing across our products before formally releasing a change.

**Comment by alexeagle** [comment]
Angular publishes snapshots only on green builds - I think it's more typical for this to be pre-release testing. We'd want to rate-limit these PRs also - if there's a new snapshot for every commit it would be too much churn.

**Comment by rarkins** [comment]
Angular snapshot builds reference for completeness: https://github.com/angular/angular/blob/master/docs/DEVELOPER.md#publishing-snapshot-builds

Unless I'm mistaken, Angular's approach of a dedicated snapshot repo is essentially "proprietary", however it seems like it would benefit a large enough audience to be worthwhile supporting this approach even if initially it means some Angular-specific logic inside Renovate. Plus I think the industry needs *something* like this so maybe it evolves into a pseudo-standard over time for cases when publish pre-releases to npmjs is not feasible or desirable.

1. We need the downstream repos to "opt-in" to being a canary repo - no problem, can be done inside `renovate` config

2. These repos would need to *test* the canary snapshot releases, so either:
   * Renovate has write permission to create/update branches for this (one per upstream package opted into), OR
   * Forking Renovate has to keep a permanently-open "not to be merged" PR from its own fork, so that the PR triggers tests to be run

3. We need a way of knowing how/where from to install the snapshot/pre-release
   * Generically, this could be npm pre-releases, e.g. `5.0.2-pre.1`
   * In Angular's case, we may need to hardcode logic at first (e.g. to `angular/core-builds`) before working out later if it can be made somewhat generic

4. We need a way to close the loop and feed back to the upstream repo (e.g. `angular/angular`) if any of the downstream canaries break
   * Raise an issue?
   * Upstream repos obviously need to be in control of which canaries they want to be notified about

5. Need to consider any other undesirable side effects and how to control them, e.g. if there are lots of commits to upstream `master` and hence lots of "pre-releases", do downstream repos want to test every time? Maybe Renovate's schedules could be used for this if it becomes a problem, e.g. only test the latest snapshot once a day and ignore them for the rest of the day.

**Comment by rarkins** [comment]
Longer term an ideal follow-on solution would be that end users (webapp builders, etc) themselves could opt-in to canary testing major dependencies like Angular, and the aggregate data used to benefit the Angular core team. Such end users would be essentially donating some build cycles + suffering a little "noise" (e,g. an extra branch, extra builds on Travis) but in exchange they would be contributing to reducing the chance that future Angular releases don't "break" them as such breaks might be detectable by aggregate statistics. Assuming people have enough automated tests for results to be useful!

**Comment by alexeagle** [comment]
Yes, that last point is exactly something Angular has offered in the past, but we didn't have a good way to do it.

/cc @igorminar @stephenfluin

**Comment by StephenFluin** [comment]
We have a small project called Community CI where we partner with teams that want to pull every commit from our master repository and send us failures from a known-good repository. Message devrel@angular.io if you are interested in setting this up. Most of the work is on your side in your CI.

Every release (beta, rc, or GA) starts its life as a commit to master, and every commit to master is validated extensively prior to merge, so by following every build from the build system, you are essentially following the same releases as Google.

**Comment by alexeagle** [comment]
Stephen - this issue is for a better automation tool to give to your
Community CI users, including ourselves.

**Comment by dzearing** [comment]
I want this for pretty much every Microsoft library, especially the far upstream ones that everything depends on.
